### PR TITLE
code_relocation: Add NOKEEP option

### PR DESCRIPTION
**Pull request by 57300**
When using the code and data relocation feature, every relocated symbol would be marked with `KEEP()` in the generated linker script. Therefore, if any input files contained unused code, then it wouldn't be discarded by the linker, even when invoked with `--gc-sections`.

This can cause unexpected bloat, or other link-time issues stemming from some symbols being discarded and others not.

On the other hand, this behavior has been present since the feature's introduction, so it should remain default for the users who rely on it.

This patch introduces support for `zephyr_code_relocate(... NOKEEP)`. This will suppress the generation of `KEEP()` statements for all symbols in a particular library or set of files.

Much like `NOCOPY`, the `NOKEEP` flag is passed to `gen_relocate_app.py` in string form. The script is now equipped to handle multiple such flags when passed from CMake as a semicolon-separated list, like so:
```
"SRAM2:NOCOPY;NOKEEP:/path/to/file1.c;/path/to/file2.c"
```
Documentation and tests are updated here as well.

### CODE REVIEW

**Review by nordicjm — CHANGES_REQUESTED** [review]
Seeing very funny behaviour with this, starting with `samples/application_development/code_relocation_nocopy` and building without NOKEEP flags, then editing the `ext_code.c` file to have a new function and building, then editing the `sram_code.c` file to have a new function and building, it seems the second build gives the same as the first build which is wrong, then the third build suddenly increases in extflash size, flash size, ram size. Adding and removing NOKEEP to each file and building again doesn't always seem to change things, in fact you can get into a state whereby changing all 3 instances from without NOKEEP to having NOKEEP yields the exact same thing, which I would assume seems to be from a state cache?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Initial looks seems ok, but the observation made by @nordicjm should be investigated.

Only major notice is that it appears there originally was a specific reason for not using `;` for anything else than the CMake list of files:
https://github.com/zephyrproject-rtos/zephyr/blob/900f06f54a26722f2d0ce6271b4618d47f852b91/cmake/modules/extensions.cmake#L1407-L1409

Therefore the following must be done:
- [x] Ensure that introducing `;` as flag list separator doesn't introduce unintended side-effects / breaks existing use-cases.
      Currently I don't see why this PR should break existing use-cases (even use-cases using genex), but I haven't looked at all existing code parts.
      Maybe a closer look at https://github.com/zephyrproject-rtos/zephyr/commit/10e85bf764aeda41f060cc971b6c892b505c813d can provide extra details.
- [x] Update referenced comment, as `;` is now used in more than the file list.

File: scripts/build/gen_relocate_app.py [hunk 1]
```diff
@@ -541,7 +560,7 @@ def create_dict_wrt_mem():
         if args.verbose:
             print("Memory region ", mem_region, " Selected for files:", file_name_list)
 
-        mem_region = "|".join((mem_region, copy_flag))
+        mem_region = "|".join((mem_region, *flag_list.split(';')))
```
[the remark this answers is not in the document]
Comment (57300) [reply scripts/build/gen_relocate_app.py]: Good idea. Done.

**Review by 57300 — COMMENTED** [review]
Thanks for your observations.

> Seeing very funny behaviour with this, starting with `samples/application_development/code_relocation_nocopy` and building without NOKEEP flags, then editing the `ext_code.c` file to have a new function and building, then editing the `sram_code.c` file to have a new function and building, it seems the second build gives the same as the first build which is wrong, then the third build suddenly increases in extflash size, flash size, ram size.

@nordicjm I tried something similar, first on `qemu_cortex_m3`:

1. Added an empty `void dummy1(void)` to `ext_code.c`, rebuilt, and saw EXTFLASH increase by 4 bytes.
2. Added an empty `void dummy2(void)` to `sram_code.c`, rebuilt, and saw RAM increase by 8 bytes.

When I repeated this on `nrf5340dk_nrf5340_cpuapp`, I saw no change in RAM size, but `dummy2` was still present in the ELF. That could be a matter of alignment or optimization.

So, I'm convinced that this is working OK. The relocation script should be re-run on every source code change, because it's set up to depend on "all" libraries:

https://github.com/zephyrproject-rtos/zephyr/blob/5fc96ff9d70f7828b5eada42de04c521cd0904e1/cmake/linker/ld/target_relocation.cmake#L35

> Adding and removing NOKEEP to each file and building again doesn't always seem to change things, in fact you can get into a state whereby changing all 3 instances from without NOKEEP to having NOKEEP yields the exact same thing, which I would assume seems to be from a state cache?

That turns out to be a pre-existing issue, but I don't mind fixing it here. Check out the second commit.

**Review by nordicjm — APPROVED** [review]
Fix works, thanks.

File: cmake/modules/extensions.cmake [hunk 2]
```diff
@@ -1392,21 +1394,25 @@ function(zephyr_code_relocate)
...
   if(CODE_REL_PHDR)
     set(CODE_REL_LOCATION "${CODE_REL_LOCATION}\ :${CODE_REL_PHDR}")
   endif()
-  # We use the "|" character to separate code relocation directives instead
-  # of using CMake lists. This way, the ";" character can be reserved for
-  # generator expression file lists.
+  # We use the "|" character to separate code relocation directives, instead of
+  # using set_property(APPEND) to produce a ";"-separated CMake list. This way,
+  # each directive can embed multiple CMake lists, representing flags and files,
+  # the latter of which can come from generator expressions.
```
[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/modules/extensions.cmake:1410]: yes, my initial conclusion was also that your changes should not have unintended side-effects, but I didn't examined the existing code in details, as we know, the devil is in the details 😉 .

Thanks for carrying out more in-depth analysis and updating the comment.

**Review by tejlmand — APPROVED** [review]
lgtm

**Pull request merged**